### PR TITLE
Application Command Support

### PR DESCRIPTION
**Pull request by alentoghostflame**
## Summary

Adds Application/Context/Slash/User/Message Command support to Nextcord, in a similar style of ext.commands without compromising future usability or maintainability of these commands.

You better squish these commits, smh.

## Checklist

- [X] If code changes were made then they have been tested.
    - [ ] I have updated the documentation to reflect the changes.
- [ ] This PR fixes an issue.
- [X] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: nextcord/application_command.py [hunk 1]
```diff
@@ -0,0 +1,1266 @@
...
+    "ApplicationSubcommand",
+    "ApplicationCommand",
+    "CommandOption",
+    "slash_command",
+    "message_command",
+    "user_command",
+)
+
+
+class InvalidCommandType(Exception):
```
[the remark this answers is not in the document]
Comment (Skelmis) [reply nextcord/application_command.py:70]: This should also inherit DiscordException or the base nextcord exception relevant to its usage.

File: nextcord/application_command.py [hunk 2]
```diff
@@ -0,0 +1,1266 @@
...
+                 description: str = MISSING,
+                 required: bool = MISSING,
+                 choices: Dict[str, Union[str, int, float]] = MISSING,
+                 channel_types: List[Type[ChannelType]] = MISSING,
+                 min_value: Union[int, float] = MISSING,
+                 max_value: Union[int, float] = MISSING,
+                 autocomplete: bool = MISSING,
+                 default: Optional[Any] = None,
+                 verify: bool = True
+                 ):
```
Comment (Skelmis) [inline nextcord/application_command.py:116]: Fix this indentation / layout to be more pep 8 inline. Deindent by 1, move self and name to under ``__init__`` and on there own line

File: nextcord/application_command.py [hunk 3]
```diff
@@ -0,0 +1,1266 @@
...
+        """Returns a payload meant for Discord for this specific Option.
+
+        Options that are not specified AND not required won't be in the returned payload.
+
+        Returns
+        -------
+        payload: :class:`dict`
+            The Discord payload for this specific Option.
+        """
+        # TODO: Figure out why pycharm is being a dingus about self.type.value being an unsolved attribute.
```
Comment (Skelmis) [inline nextcord/application_command.py:307]: Figured it out?

File: nextcord/application_command.py [hunk 4]
```diff
@@ -0,0 +1,1266 @@
...
+
+    # Data wrangling.
+
+    @property
+    def payload(self) -> dict:
+        """Verifies the content of the ApplicationCommand then constructs and returns the payload for this subcommand.
+
+        Returns
+        -------
+
```
Comment (Skelmis) [inline nextcord/application_command.py:461]: Returns what..

File: nextcord/application_command.py [hunk 5]
```diff
@@ -0,0 +1,1266 @@
...
+            for option in uncalled_options.values():
+                if option.functional_name in autocomplete_kwargs:
+                    kwargs[option.functional_name] = option.default
+            await self.invoke_autocomplete(interaction, focused_option, focused_option_value, **kwargs)
+        else:
+            raise NotImplementedError(f"{self.error_name} Autocomplete isn't handled by this type of command, how did "
+                                      f"you get here?")
+
+    async def invoke_autocomplete(self, interaction: Interaction, focused_option: CommandOption,
+                                  focused_option_value: Any, **kwargs) -> None:
```
Comment (Skelmis) [inline nextcord/application_command.py:537]: Clean up this

File: nextcord/application_command.py [hunk 6]
```diff
@@ -0,0 +1,1266 @@
...
+        elif self.type in (ApplicationCommandType.chat_input, ApplicationCommandOptionType.sub_command):
+            # Slash commands are able to have subcommands, therefore that is handled here.
+            await self.call_invoke_slash(state, interaction, option_data)
+        else:
+            # Anything that can't be handled in here should be raised for ApplicationCommand to handle.
+            # TODO: Figure out how to hide this in exception trace log, people don't need to see it.
+            raise InvalidCommandType(f"{self.type} is not a handled Application Command type.")
+
+    async def call_invoke_slash(self, state: ConnectionState, interaction: Interaction,
+                                option_data: List[Dict[str, Any]]) -> None:
```
Comment (Skelmis) [inline nextcord/application_command.py:586]: Clean this up

File: nextcord/application_command.py [hunk 7]
```diff
@@ -0,0 +1,1266 @@
...
+        Parameters
+        ----------
+        interaction: :class:`Interaction`
+            Interaction associated with the autocomplete event.
+        kwargs:
+            Keyword arguments to forward to the callback.
+
+        Returns
+        -------
+
```
Comment (Skelmis) [inline nextcord/application_command.py:632]: Why this?

File: nextcord/application_command.py [hunk 8]
```diff
@@ -0,0 +1,1266 @@
...
+                                     f"autocomplete function.")
+        if found is False:
+            raise TypeError(f"{self.error_name} kwarg {on_kwarg} not found, cannot add autocomplete function.")
+
+    def subcommand(self, **kwargs) -> Callable:
+        """Decorates a function, creating a subcommand with the given kwargs forwarded to it.
+
+        Parameters
+        ----------
+        kwargs
```
Comment (Skelmis) [inline nextcord/application_command.py:681]: kwargs, what?

File: nextcord/application_command.py [hunk 9]
```diff
@@ -0,0 +1,1266 @@
...
+        """
+        def decorator(func: Callable):
+            result = ApplicationSubcommand(callback=func, parent_command=self,
+                                           cmd_type=ApplicationCommandOptionType.sub_command, **kwargs)
```
Comment (Skelmis) [inline nextcord/application_command.py:685]: Clean this up

File: nextcord/application_command.py [hunk 10]
```diff
@@ -0,0 +1,1266 @@
...
+class ApplicationCommand(ApplicationSubcommand):
+    def __init__(
+        self,
+        callback: Optional[Callable] = MISSING,
+        cmd_type: ApplicationCommandType = MISSING,
+        name: Optional[str] = MISSING,
+        description: Optional[str] = MISSING,
+        guild_ids: Optional[Iterable[int]] = MISSING,
+        default_permission: Optional[bool] = MISSING,
+        force_global: bool = False,
```
Comment (Skelmis) [inline nextcord/application_command.py:701]: This is optional?

File: nextcord/application_command.py [hunk 11]
```diff
@@ -0,0 +1,1266 @@
...
+    def __init__(self, name: str = MISSING,
+                 description: str = MISSING,
+                 required: bool = MISSING,
+                 choices: Dict[str, Union[str, int, float]] = MISSING,
+                 channel_types: List[Type[ChannelType]] = MISSING,
+                 min_value: Union[int, float] = MISSING,
+                 max_value: Union[int, float] = MISSING,
+                 autocomplete: bool = MISSING,
+                 default: Optional[Any] = None,
+                 verify: bool = True
```
Comment (Skelmis) [inline nextcord/application_command.py:115]: This is optional?

File: nextcord/application_command.py [hunk 12]
```diff
@@ -0,0 +1,1266 @@
...
+        return self
+
+    def get_guild_payload(self, guild_id: int) -> dict:
+        """Returns a guild application command payload.
+
+        Parameters
+        ----------
+        guild_id: :class:`int`
+            Discord Guild ID to make the payload for.
+        Returns
```
Comment (Skelmis) [inline nextcord/application_command.py:809]: Newline inbetween

File: nextcord/application_command.py [hunk 13]
```diff
@@ -0,0 +1,1266 @@
...
+        """Returns a basic signature for the application command.
+
+        This signature is unique, in the sense that two application commands with the same signature cannot be
+        registered with Discord at the same time.
+
+        Parameters
+        ----------
+        guild_id: Optional[:class:`int`]
+            Integer Guild ID for the signature. For a global application command, None is used.
+        Returns
```
Comment (Skelmis) [inline nextcord/application_command.py:828]: newline

File: nextcord/application_command.py [hunk 14]
```diff
@@ -0,0 +1,1266 @@
...
+        for guild_id in self._guild_ids_to_rollout:
+            ret.add((self.name, self.type.value, guild_id))
+        return ret
+
+    # Shortcuts for ApplicationCommand creation.
+
+    @classmethod
+    def from_callback(cls, callback: Callable) -> ApplicationCommand:
+        """Returns an ApplicationCommand object created from the given callback. Overridden from ApplicationSubcommand
+        for typing purposes."""
```
Comment (Skelmis) [inline nextcord/application_command.py:860]: This docstring formatting could be cleaned up

File: nextcord/application_command.py [hunk 15]
```diff
@@ -0,0 +1,1266 @@
...
+        # if guild_id:
+        command_id = int(data["id"])
+        if guild_id := data.get("guild_id", None):
+            # self._guild_command_ids[guild_id] = command_id
+            self._command_ids[guild_id] = command_id
+            self._guild_ids.add(guild_id)
+            self.add_guild_rollout(guild_id)
+        else:
+            # self._global_command_id = command_id
+            self._command_ids[None] = command_id
```
Comment (Skelmis) [inline nextcord/application_command.py:877]: Remove the commented code if no longer needed?

File: nextcord/application_command.py [hunk 16]
```diff
@@ -0,0 +1,1266 @@
...
+
+    @property
+    def payload(self) -> List[dict]:
+        """Returns a list of Discord "Application Command Structure" payloads.
+
+        Returns
+        -------
+        payloads: List[:class:`dict`]
+            Returns a list containing the global command payload, if enabled, and payloads corresponding to every guild
+            ID specified.
```
Comment (Skelmis) [inline nextcord/application_command.py:911]: Could be formatted a bit more evenly?

File: nextcord/application_command.py [hunk 17]
```diff
@@ -0,0 +1,1266 @@
...
+            for guild_id in self.guild_ids:
+                temp = (partial_payload.copy())  # We don't need to make a deep copy as guild_id is on the top layer.
+                temp["guild_id"] = guild_id
+                ret.append(temp)
+        if self.is_global:
+            ret.append(partial_payload)
+        return ret
+
+    def check_against_raw_payload(self, raw_payload: dict, guild_id: Optional[int]) -> bool:
+        """Checks if self.payload values match with what the given raw payload has.
```
Comment (Skelmis) [inline nextcord/application_command.py:927]: ``self.payloads`` (The markdown `)

File: nextcord/application_command.py [hunk 18]
```diff
@@ -0,0 +1,1266 @@
...
+        if self.is_guild:
+            for guild_id in self.guild_ids:
+                temp = (partial_payload.copy())  # We don't need to make a deep copy as guild_id is on the top layer.
+                temp["guild_id"] = guild_id
+                ret.append(temp)
+        if self.is_global:
+            ret.append(partial_payload)
+        return ret
+
+    def check_against_raw_payload(self, raw_payload: dict, guild_id: Optional[int]) -> bool:
```
Comment (Skelmis) [inline nextcord/application_command.py:926]: ``guild_id`` is marked optional, but a guild id is required here? Default to ``None``

File: nextcord/application_command.py [hunk 19]
```diff
@@ -0,0 +1,1266 @@
...
+    def _check_against_raw_payload(self, raw_payload: dict, guild_id: Optional[int]) -> bool:
+        if guild_id:
+            cmd_payload = self.get_guild_payload(guild_id)
+            if cmd_payload["guild_id"] != int(raw_payload["guild_id"]):
+                return False
+        else:
+            cmd_payload = self.global_payload
+
+        if not check_dictionary_values(cmd_payload, raw_payload, "default_permission", "description", "type", "name"):
+            print("Failed first main check.")
```
Comment (Skelmis) [inline nextcord/application_command.py:965]: Nope. Using logging not prints.

File: nextcord/application_command.py [hunk 20]
```diff
@@ -0,0 +1,1266 @@
...
+                    # At this time, ApplicationCommand options are identical between locally-generated payloads and
+                    # payloads from Discord. If that were to change, switched to a recursive setup and manually
+                    # check_dictionary_values.
+                    if not deep_dictionary_check(cmd_option, raw_option):
+                        return False
+            if not found_correct_value:
+                return False
+        return True
+
+    def reverse_check_against_raw_payload(self, raw_payload: dict, guild_id: Optional[int]) -> bool:
```
Comment (Skelmis) [inline nextcord/application_command.py:982]: ``guild_id`` is marked optional, but a guild id is required here? Default to ``None``

File: nextcord/application_command.py [hunk 21]
```diff
@@ -0,0 +1,1266 @@
...
+        """Checks if the given raw payload values match with what self.payload has.
+
+        This doesn't make sure they are equal, and works opposite of check_against_raw_payload. This checks if all
+        key:value's inside the raw_payload also exist inside one of our payloads.
+
+        Parameters
+        ----------
+        raw_payload: :class:`dict`
+            Dictionary payload to compare against our payloads.
+        guild_id: :class:`int`
```
Comment (Skelmis) [inline nextcord/application_command.py:992]: This is optional?

File: nextcord/application_command.py [hunk 22]
```diff
@@ -0,0 +1,1266 @@
...
+        message = self._handle_resolved_message(list(interaction.data["resolved"]["messages"].values())[0])
+        await self.invoke_message(interaction, message)
+
+    async def invoke_message(self, interaction: Interaction, message: Message, **kwargs: Dict[Any, Any]) -> None:
+        if self._self_argument:
+            await self.callback(self._self_argument, interaction, message, **kwargs)
+        else:
+            await self.callback(interaction, message, **kwargs)
+
+    async def call_invoke_user(self, interaction: Interaction) -> None:
```
Comment (Skelmis) [inline nextcord/application_command.py:1100]: Document in a docstring

File: nextcord/application_command.py [hunk 23]
```diff
@@ -0,0 +1,1266 @@
...
+        user = self._handle_resolved_user(list(interaction.data["resolved"]["users"].values())[0])
+        if interaction.guild and (member := interaction.guild.get_member(user.id)):
+            await self.invoke_user(interaction, member)
+        else:
+            await self.invoke_user(interaction, user)
+
+    async def invoke_user(self, interaction: Interaction, member: Union[Member, User], **kwargs: Dict[Any, Any]) -> None:
+        """|coro|
+
+        Invokes the callback with the given interaction, member/user, and kwargs as a User Command."""
```
Comment (Skelmis) [inline nextcord/application_command.py:1113]: Document params

File: nextcord/application_command.py [hunk 24]
```diff
@@ -0,0 +1,1266 @@
...
+        if self._self_argument:
+            await self.callback(self._self_argument, interaction, member, **kwargs)
+        else:
+            await self.callback(interaction, member, **kwargs)
+
+    # Decorators.
+
+    def subcommand(self, **kwargs):
+        """Makes a function into a subcommand."""
```
Comment (Skelmis) [inline nextcord/application_command.py:1122]: Document types, what it returns n stuff

File: nextcord/application_command.py [hunk 25]
```diff
@@ -0,0 +1,1266 @@
...
+                self.children[result.name] = result
+                return result
+            return decorator
+
+
+def slash_command(name: str = MISSING,
+                  description: str = MISSING,
+                  guild_ids: Iterable[int] = MISSING,
+                  default_permission: bool = MISSING,
+                  force_global: bool = False):
```
Comment (Skelmis) [inline nextcord/application_command.py:1138]: Clean up these args inline with previous comments

Comment (Skelmis) [inline nextcord/application_command.py:1138]: This is optional?

File: nextcord/application_command.py [hunk 26]
```diff
@@ -0,0 +1,1266 @@
...
+    ----------
+    name: :class:`str`
+        Name of the command that users will see. If not set, it defaults to the name of the callback.
+    description: :class:'str'
+        Description of the command that users will see. If not set, it defaults to the bare minimum Discord allows.
+    guild_ids: Iterable[:class:`int`]
+        IDs of :class:`Guild`'s to add this command to. If unset, this will be a global command.
+    default_permission: :class:`bool`
+        If users should be able to use this command by default or not. Defaults to Discords default.
+    force_global: :class:`bool`
```
Comment (Skelmis) [inline nextcord/application_command.py:1152]: State default

File: nextcord/application_command.py [hunk 27]
```diff
@@ -0,0 +1,1266 @@
...
+        If True, will force this command to register as a global command, even if `guild_ids` is set. Will still
+        register to guilds. Has no effect if `guild_ids` are never set or added to.
+    """
+    def decorator(func: Callable) -> ApplicationCommand:
+        if isinstance(func, ApplicationCommand):
+            raise TypeError("Callback is already an ApplicationCommandRequest.")
+        app_cmd = ApplicationCommand(callback=func, cmd_type=ApplicationCommandType.chat_input, name=name,
+                                     description=description, guild_ids=guild_ids,
+                                     default_permission=default_permission, force_global=force_global)
```
Comment (Skelmis) [inline nextcord/application_command.py:1161]: beautify this

File: nextcord/application_command.py [hunk 28]
```diff
@@ -0,0 +1,1266 @@
...
+        return app_cmd
+    return decorator
+
+
+def message_command(name: str = MISSING,
+                    description: str = MISSING,
+                    guild_ids: Iterable[int] = MISSING,
+                    default_permission: bool = MISSING,
+                    force_global: bool = False):
```
Comment (Skelmis) [inline nextcord/application_command.py:1170]: beautify this

Comment (Skelmis) [inline nextcord/application_command.py:1170]: ``force_global`` is optional?

File: nextcord/application_command.py [hunk 29]
```diff
@@ -0,0 +1,1266 @@
...
+    ----------
+    name: :class:`str`
+        Name of the command that users will see. If not set, it defaults to the name of the callback.
+    description: :class:'str'
+        Description of the command that users will see. If not set, it defaults to the bare minimum Discord allows.
+    guild_ids: Iterable[:class:`int`]
+        IDs of :class:`Guild`'s to add this command to. If unset, this will be a global command.
+    default_permission: :class:`bool`
+        If users should be able to use this command by default or not. Defaults to Discords default.
+    force_global: :class:`bool`
```
Comment (Skelmis) [inline nextcord/application_command.py:1184]: Denote default

File: nextcord/application_command.py [hunk 30]
```diff
@@ -0,0 +1,1266 @@
...
+        If True, will force this command to register as a global command, even if `guild_ids` is set. Will still
+        register to guilds. Has no effect if `guild_ids` are never set or added to.
+    """
+    def decorator(func: Callable) -> ApplicationCommand:
+        if isinstance(func, ApplicationCommand):
+            raise TypeError("Callback is already an ApplicationCommandRequest.")
+        app_cmd = ApplicationCommand(callback=func, cmd_type=ApplicationCommandType.message, name=name,
+                                     description=description, guild_ids=guild_ids,
+                                     default_permission=default_permission, force_global=force_global)
```
Comment (Skelmis) [inline nextcord/application_command.py:1193]: beautify

File: nextcord/application_command.py [hunk 31]
```diff
@@ -0,0 +1,1266 @@
...
+        return app_cmd
+    return decorator
+
+
+def user_command(name: str = MISSING,
+                 description: str = MISSING,
+                 guild_ids: Iterable[int] = MISSING,
+                 default_permission: bool = MISSING,
+                 force_global: bool = False):
```
Comment (Skelmis) [inline nextcord/application_command.py:1202]: Beautify and this is optional?

File: nextcord/application_command.py [hunk 32]
```diff
@@ -0,0 +1,1266 @@
...
+    force_global: :class:`bool`
+        If True, will force this command to register as a global command, even if `guild_ids` is set. Will still
+        register to guilds. Has no effect if `guild_ids` are never set or added to.
+    """
+    def decorator(func: Callable) -> ApplicationCommand:
+        if isinstance(func, ApplicationCommand):
+            raise TypeError("Callback is already an ApplicationCommandRequest.")
+        app_cmd = ApplicationCommand(callback=func, cmd_type=ApplicationCommandType.user, name=name,
+                                     description=description, guild_ids=guild_ids,
+                                     default_permission=default_permission, force_global=force_global)
```
Comment (Skelmis) [inline nextcord/application_command.py:1225]: beautify

File: nextcord/application_command.py [hunk 33]
```diff
@@ -0,0 +1,1266 @@
...
+        return app_cmd
+    return decorator
+
+
+def check_dictionary_values(dict1, dict2, *keywords) -> bool:
```
Comment (Skelmis) [inline nextcord/application_command.py:1230]: Type these as dicts?

File: nextcord/client.py [hunk 34]
```diff
@@ -29,46 +29,51 @@
...
 from .sticker import GuildSticker, StandardSticker, StickerPack, _sticker_factory
+from .template import Template
+from .threads import Thread
+from .ui.view import View
+from .user import User, ClientUser
+from .utils import MISSING
+from .voice_client import VoiceClient
+from .webhook import Webhook
+from .widget import Widget
+
```
Comment (Skelmis) [inline nextcord/client.py:66]: Github makes this look messy, but maybe run `isort` over these?

File: nextcord/client.py [hunk 35]
```diff
@@ -193,6 +199,28 @@ class Client:
...
+        locally added command.
+
+    rollout_register_new: :class:`bool`
+        Whether during the application command rollout to register new application commands that were added locally but
+        not found on Discord.
+
+    rollout_update_known: :class:`bool`
+        Whether during the application command rollout to update known applications that share the same signature but
+        don't quite match what is registered on Discord.
+
```
Comment (Skelmis) [inline nextcord/client.py:222]: These are optional? 
Denote defaults

File: nextcord/client.py [hunk 36]
```diff
@@ -1643,3 +1677,241 @@ def persistent_views(self) -> Sequence[View]:
         .. versionadded:: 2.0
         """
         return self._connection.persistent_views
+
+    async def on_interaction(self, interaction: Interaction):
+        await self.process_application_commands(interaction)
+
+    async def process_application_commands(self, interaction: Interaction) -> None:
```
Comment (Skelmis) [inline nextcord/client.py:1684]: Document params

File: nextcord/client.py [hunk 37]
```diff
@@ -1643,3 +1677,241 @@ def persistent_views(self) -> Sequence[View]:
...
+                _log.info(f"nextcord.Client: Calling your application command now {app_cmd.name}")
+                await app_cmd.call_from_interaction(interaction)
+            elif self._lazy_load_commands:
+                _log.info(f"nextcord.Client: Interaction command not found, attempting to lazy load.")
+                _log.debug(f"nextcord.Client: {interaction.data}")
+                response_signature = (interaction.data["name"], int(interaction.data['type']), interaction.guild_id)
+                _log.debug(f"nextcord.Client: {response_signature}")
+                if app_cmd := self._connection.get_application_command_from_signature(interaction.data["name"],
+                                                                                      int(interaction.data['type']),
+                                                                                      interaction.guild_id):
```
Comment (Skelmis) [inline nextcord/client.py:1700]: This can be beautified

File: nextcord/client.py [hunk 38]
```diff
@@ -1643,3 +1677,241 @@ def persistent_views(self) -> Sequence[View]:
...
+                await app_cmd.call_autocomplete_from_interaction(interaction)
+            else:
+                raise ValueError(f"Received autocomplete interaction for {interaction.data['name']} but command isn't "
+                                 f"found/associated!")
+
+    def get_application_command(self, command_id: int) -> Optional[ApplicationCommand]:
+        return self._connection.get_application_command(command_id)
+
+    def add_application_command(self, command: ApplicationCommand, overwrite: bool = False,
+                                use_rollout: bool = False) -> None:
```
Comment (Skelmis) [inline nextcord/client.py:1722]: Beautify this

``use_rollout`` is optional?

File: nextcord/client.py [hunk 39]
```diff
@@ -1643,3 +1677,241 @@ def persistent_views(self) -> Sequence[View]:
...
+        """Adds an ApplicationCommand object to the client for use.
+
+        Parameters
+        ----------
+        command: :class:`ApplicationCommand`
+            Command to add to the client for usage.
+        overwrite: :class:`bool`
+            If to overwrite any existing commands that would conflict with this one.
+        use_rollout: :class:`bool`
```
Comment (Skelmis) [inline nextcord/client.py:1731]: denote default

File: nextcord/client.py [hunk 40]
```diff
@@ -1643,3 +1677,241 @@ def persistent_views(self) -> Sequence[View]:
...
+            If to apply the rollout signatures instead of existing ones.
+        """
+        self._connection.add_application_command(command, overwrite=overwrite, use_rollout=use_rollout)
+
+    async def deploy_application_commands(self, data: Optional[List[dict]] = None, associate_known: bool = True,
+                                          delete_unknown: bool = True, update_known: bool = True) -> None:
```
Comment (Skelmis) [inline nextcord/client.py:1737]: beautify and document

File: nextcord/client.py [hunk 41]
```diff
@@ -1643,3 +1677,241 @@ def persistent_views(self) -> Sequence[View]:
...
+        """Updates and associates recognizable global application commands, and deletes unknown ones."""
+        # TODO: Look, I need to get this out of the door. Properish documentation of this is in ConnectionState, and
+        #  this could be an easy "I want to contribute to Nextcord in a meaningful way" PR.
+        await self._connection.deploy_application_commands(data=data, guild_id=None, associate_known=associate_known,
+                                                           delete_unknown=delete_unknown, update_known=update_known)
+
+    async def delete_unknown_application_commands(self, data: Optional[List[dict]] = None) -> None:
+        """Deletes unknown global commands."""
```
Comment (Skelmis) [inline nextcord/client.py:1745]: Document params

File: nextcord/client.py [hunk 42]
```diff
@@ -1643,3 +1677,241 @@ def persistent_views(self) -> Sequence[View]:
...
+
+    async def on_connect(self) -> None:
+        await self.rollout_global_application_commands()
+
+    async def rollout_global_application_commands(self) -> None:
+        self.add_all_cog_commands()
+        global_payload = await self.http.get_global_commands(self.application_id)
+        await self.deploy_application_commands(data=global_payload, associate_known=self._rollout_associate_known,
+                                               delete_unknown=self._rollout_delete_unknown,
+                                               update_known=self._rollout_update_known)
```
Comment (Skelmis) [inline nextcord/client.py:1803]: beautify

File: nextcord/client.py [hunk 43]
```diff
@@ -1643,3 +1677,241 @@ def persistent_views(self) -> Sequence[View]:
...
+                    self.add_application_command(cmd, use_rollout=True)
+
+    def add_cog(self, cog: ClientCog) -> None:
+        self._client_cogs.add(cog)
+
+    def user_command(self, name: str = MISSING,
+                     description: str = MISSING,
+                     guild_ids: Iterable[int] = MISSING,
+                     default_permission: bool = MISSING,
+                     force_global: bool = False):
```
Comment (Skelmis) [inline nextcord/client.py:1836]: beautify and this is optional?

File: nextcord/client.py [hunk 44]
```diff
@@ -1643,3 +1677,241 @@ def persistent_views(self) -> Sequence[View]:
...
+            self.add_application_command(result)
+            return result
+
+        return decorator
+
+    def message_command(self, name: str = MISSING,
+                        description: str = MISSING,
+                        guild_ids: Iterable[int] = MISSING,
+                        default_permission: bool = MISSING,
+                        force_global: bool = False):
```
Comment (Skelmis) [inline nextcord/client.py:1865]: beautify and this is optional?

File: nextcord/client.py [hunk 45]
```diff
@@ -1643,3 +1677,241 @@ def persistent_views(self) -> Sequence[View]:
...
+            self.add_application_command(result)
+            return result
+
+        return decorator
+
+    def slash_command(self, name: str = MISSING,
+                      description: str = MISSING,
+                      guild_ids: Iterable[int] = MISSING,
+                      default_permission: bool = MISSING,
+                      force_global: bool = False):
```
Comment (Skelmis) [inline nextcord/client.py:1894]: beautify and this is optional?

File: nextcord/guild.py [hunk 46]
```diff
@@ -2966,3 +2969,34 @@ async def change_voice_state(
         ws = self._state._get_websocket(self.id)
         channel_id = channel.id if channel else None
         await ws.voice_state(self.id, channel_id, self_mute, self_deaf)
+
+    def add_application_command(self, app_cmd: ApplicationCommand, overwrite: bool = False,
+                                use_rollout: bool = False) -> None:
```
Comment (Skelmis) [inline nextcord/guild.py:2974]: beautify

File: nextcord/guild.py [hunk 47]
```diff
@@ -2966,3 +2969,34 @@ async def change_voice_state(
...
+        app_cmd.add_guild_rollout(self.id)
+        self._state.add_application_command(app_cmd, overwrite=overwrite, use_rollout=use_rollout)
+
+    async def deploy_application_commands(self, data: Optional[List[dict]] = None, associate_known: bool = True,
+                                          delete_unknown: bool = True, update_known: bool = True):
+        await self._state.deploy_application_commands(data=data, guild_id=self.id, associate_known=associate_known,
+                                                      delete_unknown=delete_unknown, update_known=update_known)
```
Comment (Skelmis) [inline nextcord/guild.py:2981]: beautify this method

File: nextcord/state.py [hunk 48]
```diff
@@ -474,6 +487,240 @@ def _get_guild_channel(self, data: MessagePayload) -> Tuple[Union[Channel, Threa
...
+
+    def get_application_command(self, command_id: int) -> Optional[ApplicationCommand]:
+        return self._application_command_ids.get(command_id, None)
+
+    def get_application_command_from_signature(self, name: str, cmd_type: int,
+                                               guild_id: Optional[int]) -> Optional[ApplicationCommand]:
+        return self._application_command_signatures.get((name, cmd_type, guild_id), None)
+
+    def add_application_command(self, command: ApplicationCommand, overwrite: bool = False,
+                                use_rollout: bool = False) -> None:
```
Comment (Skelmis) [inline nextcord/state.py:502]: beautify these methods

File: nextcord/state.py [hunk 49]
```diff
@@ -474,6 +487,240 @@ def _get_guild_channel(self, data: MessagePayload) -> Tuple[Union[Channel, Threa
...
+        self._application_commands.remove(command)
+
+    def add_all_rollout_signatures(self) -> None:
+        """This adds all command signatures for rollouts to the signature cache."""
+        for command in self._application_commands:
+            self.add_application_command(command, use_rollout=True)
+
+    async def deploy_application_commands(self, data: Optional[List[dict]] = None, guild_id: Optional[int] = None,
+                                          associate_known: bool = True, delete_unknown: bool = True,
+                                          update_known: bool = True) -> None:
```
Comment (Skelmis) [inline nextcord/state.py:537]: beautify and this is optional? They all have defaults

Comment (Skelmis) [reply nextcord/state.py:537]: I won't make more, but can you do this for all methods you've add ere?

File: nextcord/application_command.py (hunk 3 above)
Comment (alentoghostflame) [reply nextcord/application_command.py:307]: Nope!

File: nextcord/application_command.py (hunk 11 above)
Comment (alentoghostflame) [reply nextcord/application_command.py:115]: It's optional in the sense that it can be None, unless you think Any is enough?

Comment (veni-vidi-code) [reply nextcord/application_command.py]: Shouldnt optional be implemented using MISSING then?

Comment (veni-vidi-code) [reply nextcord/application_command.py]: but Any would allow None anyways

File: nextcord/errors.py [hunk 50]
```diff
@@ -64,6 +65,11 @@ class DiscordException(Exception):
     pass
 
 
+class InvalidCommandType(DiscordException):
```
Comment (Skelmis) [inline nextcord/errors.py:68]: Put this in the docs

File: nextcord/application_command.py [hunk 51]
```diff
@@ -0,0 +1,1361 @@
...
+            Any additional keyword arguments to pass to the callback.
+        """
+        if self._self_argument:
+            await self.callback(self._self_argument, interaction, member, **kwargs)
+        else:
+            await self.callback(interaction, member, **kwargs)
+
+    # Decorators.
+
+    def subcommand(self, name: str = MISSING, description: str = MISSING) -> Callable:
```
Comment (VincentRPS) [inline nextcord/application_command.py:1177]: maybe rename this to sub_command for consistency with the other decorators?

File: nextcord/enums.py [hunk 52]
```diff
@@ -528,6 +530,7 @@ class InteractionType(Enum):
     ping = 1
     application_command = 2
     component = 3
+    application_command_autocomplete = 4
```
Comment (VincentRPS) [inline nextcord/enums.py:533]: put this on top of component

**Review by VincentRPS — CHANGES_REQUESTED** [review]
my internet won't let me review client & state so i'll do that later

File: nextcord/application_command.py (hunk 51 above)
Comment (alentoghostflame) [reply nextcord/application_command.py:1177]: A small, yet large change. I want others to okay this before going ahead with it.

**Pull request merged**